src/EditorFeatures/Core/Implementation/CommentSelection/ToggleLineCommentCommandHandler.cs

```diff
@@ -0,0 +1,191 @@
+            : base(undoHistoryRegistry, editorOperationsFactoryService)
+        {
+        }
+
+        public VSCommanding.CommandState GetCommandState(ToggleLineCommentCommandArgs args)
+        {
+            if (Workspace.TryGetWorkspace(args.SubjectBuffer.AsTextContainer(), out var workspace))
+            {
+                var experimentationService = workspace.Services.GetRequiredService<IExperimentationService>();
+                if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleLineComment))
+                {
+                    return VSCommanding.CommandState.Unspecified;
+                }
+            }
+            return GetCommandState(args.SubjectBuffer);
+        }
+
+        public bool ExecuteCommand(ToggleLineCommentCommandArgs args, CommandExecutionContext context)
+            => ExecuteCommand(args.TextView, args.SubjectBuffer, ValueTuple.Create(), context);
+
+        public override string DisplayName => EditorFeaturesResources.Toggle_Line_Comment;
+
+        protected override string GetTitle(ValueTuple command) => EditorFeaturesResources.Toggle_Line_Comment;
+
+        protected override string GetMessage(ValueTuple command) => EditorFeaturesResources.Toggling_line_comment;
+
+        internal async override Task<CommentSelectionResult> CollectEditsAsync(Document document, ICommentSelectionService service,
+            ITextBuffer subjectBuffer, NormalizedSnapshotSpanCollection selectedSpans, ValueTuple command, CancellationToken cancellationToken)
+        {
+            var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleLineComment))
+            {
+                return s_emptyCommentSelectionResult;
+            }
+
+            var commentInfo = await service.GetInfoAsync(document, selectedSpans.First().Span.ToTextSpan(), cancellationToken).ConfigureAwait(false);
+            if (commentInfo.SupportsSingleLineComment)
+            {
+                return ToggleLineComment(commentInfo, selectedSpans);
+            }
+
+            return s_emptyCommentSelectionResult;
+        }
+
+        private CommentSelectionResult ToggleLineComment(CommentSelectionInfo commentInfo,
+            NormalizedSnapshotSpanCollection selectedSpans)
+        {
+            var textChanges = ArrayBuilder<TextChange>.GetInstance();
+            var trackingSpans = ArrayBuilder<CommentTrackingSpan>.GetInstance();
+
+            var linesInSelections = selectedSpans.ToDictionary(
+                span => span,
+                span => GetLinesFromSelectedSpan(span).ToImmutableArray());
+
+            Operation operation;
+            // If all the selections are fully commented, uncomment.
+            if (!linesInSelections.Values.Where(lines => SelectionHasUncommentedLines(lines, commentInfo)).Any())
```

Comment:
have the ! with the `if/else` seems unnecessary.  can you just remove it and flip the bodies to make it clearer?